# Attack 3 - Readiness Probe Flood: Comparison

Loads baseline and attack3 CSV exports and compares them (volume, rates, deltas).
See `attacks/attack3-readiness-probe-flood.md` for the attack description and deployment manifests.

- Baseline capture: `exports/baseline/`
- Attack capture: `exports/attack3-readiness-probe-flood/`
- Comparison logic: `compare_logs.py`

In [1]:
from pathlib import Path
from IPython.display import display

ATTACK_DIR = Path("../exports/attack3-readiness-probe")
ATTACK_SUFFIX = "attack3"

%run -i compare_logs.py

print("=== Resume ===")
display(summary)

print("\n=== Events per second (stats) ===")
display(eps_stats)

print("\n=== Resource x Verb ===")
display(resource_x_verb)

for field, df in results.items():
    print(f"\n=== {field} ===")
    display(df)

        métrique  baseline  attack  variation
0   total_events      2089   21243      19154
1  patch_on_pods         0    3484       3484
=== Resume ===


,métrique,baseline,attack,variation
0,total_events,2089,21243,19154
1,patch_on_pods,0,3484,3484



=== Events per second (stats) ===


,métrique,baseline,attack,variation
0,max_events_par_sec,25.0,177.00,152.00
1,moyenne_events_par_sec,7.6,70.26,62.66



=== Resource x Verb ===


,resource,verb,count_baseline,count_attack,variation
0,deployments,update,0,5142,5142
1,replicasets,update,0,5141,5141
2,pods,get,0,3546,3546
3,pods,patch,0,3484,3484
4,events,patch,0,1707,1707
...,...,...,...,...,...
70,ippools,watch,6,0,-6
71,volumeattachments,watch,6,0,-6
72,networkpolicies,watch,9,0,-9
73,services,watch,24,15,-9



=== verb ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,update,710,10944,142.0,2188.8,10234,2046.8
1,patch,3,5195,0.6,1039.0,5192,1038.4
2,get,1063,4613,212.6,922.6,3550,710.0
3,watch,309,332,61.8,66.4,23,4.6
4,create,0,146,0.0,29.2,146,29.2
5,list,4,13,0.8,2.6,9,1.8



=== objectRef_resource ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,pods,12,7090,2.4,1418.0,7078,1415.6
1,deployments,3,5145,0.6,1029.0,5142,1028.4
2,replicasets,6,5144,1.2,1028.8,5138,1027.6
3,events,0,1853,0.0,370.6,1853,370.6
4,leases,713,667,142.6,133.4,-46,-9.2
...,...,...,...,...,...,...,...
57,networkpolicies,9,0,1.8,0.0,-9,-1.8
58,poddisruptionbudgets,3,0,0.6,0.0,-3,-0.6
59,ippools,6,0,1.2,0.0,-6,-1.2
60,kubecontrollersconfigurations,3,0,0.6,0.0,-3,-0.6



=== objectRef_namespace ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,attacker-ns,0,19208,0.0,3841.6,19208,3841.6
1,kube-system,643,606,128.6,121.2,-37,-7.4
2,default,116,112,23.2,22.4,-4,-0.8
3,kube-node-lease,88,82,17.6,16.4,-6,-1.2



=== objectRef_subresource ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,status,3,13771,0.6,2754.2,13768,2753.6



=== responseStatus_code ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,200,990,4531,198.0,906.2,3541,708.2



=== authorization_decision ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,allow,990,4531,198.0,906.2,3541,708.2



=== authorization_reason ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,"RBAC: allowed by ClusterRoleBinding ""system:co...",0,1739,0.0,347.8,1739,347.8
1,"RBAC: allowed by ClusterRoleBinding ""system:co...",0,1730,0.0,346.0,1730,346.0
2,"RBAC: allowed by ClusterRoleBinding ""system:pu...",330,303,66.0,60.6,-27,-5.4
3,"RBAC: allowed by ClusterRoleBinding ""system:ku...",236,220,47.2,44.0,-16,-3.2
4,"RBAC: allowed by ClusterRoleBinding ""system:ku...",161,154,32.2,30.8,-7,-1.4
5,"RBAC: allowed by ClusterRoleBinding ""system:di...",65,47,13.0,9.4,-18,-3.6
6,"RBAC: allowed by ClusterRoleBinding ""calico-ku...",30,30,6.0,6.0,0,0.0
7,"RBAC: allowed by ClusterRoleBinding ""kubeadm:c...",0,21,0.0,4.2,21,4.2
8,"RBAC: allowed by ClusterRoleBinding ""system:pu...",5,19,1.0,3.8,14,2.8
9,"RBAC: allowed by ClusterRoleBinding ""kubeadm:n...",14,10,2.8,2.0,-4,-0.8



=== sourceIPs ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,10.10.10.10,1572,11823,314.4,2364.6,10251,2050.2
1,10.10.10.11,83,8990,16.6,1798.0,8907,1781.4
2,::1,210,213,42.0,42.6,3,0.6
3,192.168.62.142,0,112,0.0,22.4,112,22.4
4,10.10.10.12,89,99,17.8,19.8,10,2.0
5,192.168.62.167,0,6,0.0,1.2,6,1.2
6,10.0.2.15,6,0,1.2,0.0,-6,-1.2
7,192.168.62.161,120,0,24.0,0.0,-120,-24.0
8,192.168.62.160,9,0,1.8,0.0,-9,-1.8



=== user_groups ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,system:authenticated,1429,20636,285.8,4127.2,19207,3841.4
1,system:serviceaccounts:kube-system,251,10492,50.2,2098.4,10241,2048.2
2,system:serviceaccounts,251,10492,50.2,2098.4,10241,2048.2
3,system:nodes,217,9184,43.4,1836.8,8967,1793.4
4,system:unauthenticated,660,607,132.0,121.4,-53,-10.6
5,system:masters,210,213,42.0,42.6,3,0.6
6,kubeadm:cluster-admins,0,42,0.0,8.4,42,8.4



=== user_username ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,system:node:k8s-w1,71,8984,14.2,1796.8,8913,1782.6
1,system:serviceaccount:kube-system:deployment-c...,0,5142,0.0,1028.4,5142,1028.4
2,system:serviceaccount:kube-system:replicaset-c...,0,5141,0.0,1028.2,5141,1028.2
3,system:anonymous,660,607,132.0,121.4,-53,-10.6
4,system:kube-controller-manager,428,399,85.6,79.8,-29,-5.8
5,system:kube-scheduler,323,306,64.6,61.2,-17,-3.4
6,system:apiserver,210,213,42.0,42.6,3,0.6
7,system:serviceaccount:kube-system:calico-kube-...,120,112,24.0,22.4,-8,-1.6
8,system:node:k8s-cp,72,102,14.4,20.4,30,6.0
9,system:node:k8s-w2,74,98,14.8,19.6,24,4.8



=== userAgent ===


,category,count_baseline,count_attack,taux_baseline_par_min,taux_attack_par_min,variation_absolue,variation_taux_par_min
0,kubelet/v1.30.14 (linux/amd64) kubernetes/9e18483,217,9184,43.4,1836.8,8967,1793.4
1,kube-controller-manager/v1.30.14 (linux/amd64)...,0,5142,0.0,1028.4,5142,1028.4
2,kube-controller-manager/v1.30.14 (linux/amd64)...,0,5141,0.0,1028.2,5141,1028.2
3,kube-probe/1.30,660,607,132.0,121.4,-53,-10.6
4,kube-controller-manager/v1.30.14 (linux/amd64)...,296,276,59.2,55.2,-20,-4.0
5,kube-scheduler/v1.30.14 (linux/amd64) kubernet...,296,276,59.2,55.2,-20,-4.0
6,kube-apiserver/v1.30.14 (linux/amd64) kubernet...,210,213,42.0,42.6,3,0.6
7,kube-controller-manager/v1.30.14 (linux/amd64)...,96,96,19.2,19.2,0,0.0
8,kube-controllers/v0.0.0 (linux/amd64) kubernet...,60,57,12.0,11.4,-3,-0.6
9,Go-http-client/2.0,72,55,14.4,11.0,-17,-3.4


## Results summary

| Metric | Baseline | Attack | Variation | Factor |
|---|---|---|---|---|
| Total audit events | 2089 | 21243 | +19154 | x10.2 |
| Events/sec (average) | 7.6 | 70.26 | +62.66 | x9.2 |
| Events/sec (peak) | 25 | 177 | +152 | x7.1 |
| `objectRef.subresource: status` | 3 | 13771 | +13768 | x4590 |
| `authorization.decision: allow` | 990 | 4531 | +3541 | x4.6 |

| Controller account (`userAgent`) | Baseline | Attack |
|---|---|---|
| deployment-controller | 0 | 5142 |
| replicaset-controller | 0 | 5141 |

| Resource | Verb | Baseline | Attack |
|---|---|---|---|
| pods | get | 0 | 3546 |
| pods | patch | 0 | 3484 |
| deployments | update | 0 | 5142 |
| replicasets | update | 0 | 5141 |
| events | patch | 0 | 1707 |

| Namespace | Baseline | Attack |
|---|---|---|
| attacker-ns | 0 | 19208 |

---

## Analysis of results

The breakdown by `(resource, verb)` shows a single, clean dependency chain being exercised: `pods` (patch +3484, get +3546) at the source, propagating to `replicasets` (update +5141) and then `deployments` (update +5142). `endpoints` and `endpointslices` stay essentially flat (64 to 62 and 70 to 68 respectively), confirming that only the pod-replicaset-deployment propagation path is active in this capture.

The `objectRef.subresource: status` field shows a sharp shift: from 3 to 13771 occurrences (x4590), reaching roughly 65% of total attack volume. This is consistent with one hundred independently scheduled pods each flipping readiness every second: nearly every event generated by the attack is a state update, not a structural change to the cluster.

The two controller accounts `deployment-controller` and `replicaset-controller` are entirely silent in the baseline (0 events each) and jointly responsible for 10283 events during the attack, roughly 48% of the total volume. This matches the expectation that a readiness flip reaching the API server fans out into two independent reconciliation loops (ReplicaSet status, then Deployment status) for every single pod, which explains why their combined volume is close in magnitude to the `pods patch`/`get` volume generated directly by the kubelet.

`verb: update` alone accounts for 10944 events (up from 710 in the baseline, x15.4), the largest single-verb contributor to the attack, consistent with the two controllers above each performing one `update` per pod per reconciliation cycle. `patch` goes from 3 to 5195 (x1732), matching the kubelet's periodic `pods/status` patches driven by the readiness probe.

`authorization.decision` stays at 100% `allow` (990 to 4531), and the `authorization.reason` breakdown confirms the attack uses only default RBAC grants: the `ClusterRoleBinding` entries for `system:controller:deployment-controller` and `system:controller:replicaset-controller` appear at high volume (0 to 1739 and 0 to 1730) without any new or unusual role being invoked.

---

## Detectability analysis

**1. Appearance of service accounts absent from the baseline.** `deployment-controller` and `replicaset-controller` are silent in the baseline and jointly generate 10283 events during the attack. A system account with zero baseline activity suddenly producing thousands of events per observation window is a strong, low false positive indicator.

**2. Concentration on `objectRef.subresource: status`.** The jump from 0.1% to roughly 65% of total volume on the `status` subresource is a very sharp signal. A threshold rule on this ratio would catch this attack with a wide margin.

**3. Volume and rate on `update`/`patch` verbs.** `update` alone grows by a factor of 15.4 and `patch` by a factor of 1732. A rule combining verb distribution with absolute event rate would be more robust than relying on total volume alone, since it ties the alert to the specific mechanism (status churn) rather than a generic traffic spike.

### Detectability limitations to note

- **RBAC decision always `allow`**: a detector based only on authorization denials would miss this attack entirely. Detection must rely on volume and verb distribution, not on the legitimacy of the calls.
- **Namespace confinement**: the entire signal sits in `attacker-ns` in this capture. An attacker spreading the same pattern across several namespaces, or inside a namespace already carrying legitimate churn, would require detection to fall back on the subresource and controller-account indicators rather than namespace filtering.
- **Narrow propagation footprint**: because no Service is attached to these pods, the `endpoints`/`endpointslices` indicators stay flat here. A detection rule relying specifically on endpoint/endpointslice churn would miss this pattern, which argues for keeping the `status` subresource ratio and the deployment/replicaset controller accounts as the primary indicators for this mechanism.